lib/linter/linter.js

```diff
@@ -1308,9 +1308,9 @@ class Linter {
                 return [];
             }
 
-            // Resolve configuration again if the file extension was changed.
-            if (configForRecursive && path.extname(blockName) !== originalExtname) {
-                debug("Resolving configuration again because the file extension was changed.");
+            // Resolve configuration again if the file content or extension was changed.
+            if (configForRecursive && (text !== blockText || path.extname(blockName) !== originalExtname)) {
+                debug("Resolving configuration again because the file content or extension was changed.");
```

Comment:
This is potentially a breaking change. The current behavior when `preprocess()` returns the same extension isn't well documented, but maybe some processors don't expect that they'll be run again on a part of content that originates from the same file. This also changes how existing user configurations will be interpreted (#14207).

Reply:
Is there any example that this change will break some processors? I can't imagine that.

And It's just the correct behavior.

Reply:
And by referencing my original issue, I don't understand what will change, I tried this change locally, rules of `*.foo/*`(including `*.foo/*.foo`) take higher priority than `*.foo`, it is very expected to me.

Reply:
I agree with you that the current behavior looks unexpected, but we can't really know all third-party processors to tell if this will break something. I marked this PR as "breaking" until we figure out whether the original issue is a bug or just an undocumented exception in the behavior, and estimate the impact of a fix if it's a bug.

Reply:
Then how do you figure out whether it's a bug?

Reply:
@mdjermanovic Any news?

Reply:
<pre><code># Outer

````md
# Middle

```js
console.log('Inner')
```
````
</code></pre>

To confirm my understanding, currently, if we have the above `README.md` file which itself contains `md` code blocks, we would not re-run the `md` processor on the nested code block today but would after this change?

If that's correct, I'd look at this as kind of making two changes:

1. We would re-evaluate the config for the code block, so `README.md` and `README.md/0_0.md` could have different configs. That's the most specific fix to #14207 so that inner code blocks get `overrides` applied.
2. As a side effect of this change's implementation, we would start running processors recursively, which would then extract the `js` code block to be linted as `README.md/0_0.md/0_0.js`. https://github.com/eslint/eslint-plugin-markdown/issues/181 indicates that at least the Markdown plugin wasn't designed to be run recursively and would break if it were.

The config improvments RFC [accounted for nested processors](https://github.com/eslint/rfcs/tree/master/designs/2018-processors-improvements#virtual-filenames-for-code-blocks) but did not specifically address recursive processors:

> CLIEngine will call getConfigForFile() on this filename to determine the correct configuration for the code block, which includes whether another processor should be run on the code block (matched by file extension).

However, I'm leaning toward that second part being a bug as well. Previously, a `.md` file containing an `html` code block that contains a `md` code block would end up running the `md` processor recursively in today's implementation because the extension check would be different each time (`md` -> `html` -> `md`). This change just removes the requirement for the intermediate code block with a different extension, which I infer was not intended to be a requirement.

I see part 2 as a bug fix that will break at least one processor and possibly others.

Reply:
> I see part 2 as a bug fix that will break at least one processor and possibly others.

Is it still an open question whether we will fix only 1, or both 1 and 2?

Reply:
As @btmills mentioned, `foo.md/0_0.markdown/0_0.js` is linted now, so 2) should be fixed too.

Reply:
Yes, I noticed that `eslint-plugin-markdown` has been fixed and released, I'm just concerned about the "possibly others" part.

Reply:
That issue is not caused by this fix at all, and `md -> html -> js`, so I don't think we should concern about that part.

Reply:
I agree. If other processors have the same implementation bug as `eslint-plug-in-markdown`, it's already possible to repro with a different syntax in between, so this is making it slightly easier to encounter in the worst case.